extensions-contrib/prometheus-emitter/src/main/java/org/apache/druid/emitter/prometheus/PrometheusEmitter.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.emitter.prometheus;
+
+
+import com.google.common.collect.ImmutableMap;
+import io.prometheus.client.Counter;
+import io.prometheus.client.Gauge;
+import io.prometheus.client.Histogram;
+import io.prometheus.client.exporter.HTTPServer;
+import io.prometheus.client.exporter.PushGateway;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.java.util.emitter.core.Emitter;
+import org.apache.druid.java.util.emitter.core.Event;
+import org.apache.druid.java.util.emitter.service.ServiceMetricEvent;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+/**
+ *
+ */
+public class PrometheusEmitter implements Emitter
+{
+
+  private static final Logger log = new Logger(PrometheusEmitter.class);
+  private final Metrics metrics;
+  private final PrometheusEmitterConfig config;
+  private final PrometheusEmitterConfig.Strategy strategy;
+  private final Pattern pattern = Pattern.compile("[^a-zA-Z0-9_][^a-zA-Z0-9_]*");
+
+  private HTTPServer server;
+  private PushGateway pushGateway;
+  private String identifier;
+
+  static PrometheusEmitter of(PrometheusEmitterConfig config)
+  {
+    return new PrometheusEmitter(config);
+  }
+
+  public PrometheusEmitter(PrometheusEmitterConfig config)
+  {
+    this.config = config;
+    this.strategy = config.getStrategy();
+    metrics = new Metrics(config.getNamespace(), config.getDimensionMapPath());
+  }
+
+
+  @Override
+  public void start()
```

Comment:
we should schedule a task to push updates periodically when the strategy is set to `pushgateway`

Reply:
Added, every 5min sounds reasonable?

Reply:
Sorry I missed this - I think the scheduled executor may not be necessary. Main reason we've added strategy `pushgateway` is for things that are potentially too short-lived to be scraped by prometheus. Things that are living long enough to be pushing every 5 minutes are likely not "task" based, and may be better fit for normal scraping.  I lean toward keeping things simple, and pushing once at close seems sufficient.

Reply:
Given the only metric pushed by peon is "last pushed timestamp", I think it's valid to remove the scheduled task. Removed.